File: generated-code/csharp/codecraft/TcpReadWrite/Model/BuildAction.cs
namespace TransGenTest.Model
{
    /// <summary>
    /// Build action
    /// </summary>
    public struct BuildAction
    {
        /// <summary>
        /// Type of an entity to build
        /// </summary>
        public Model.EntityType EntityType { get; set; }
        /// <summary>
        /// Desired position of new entity
        /// </summary>
        public Vec2Int Position { get; set; }
    
        public BuildAction(Model.EntityType entityType, Vec2Int position)
        {
            this.EntityType = entityType;
            this.Position = position;
        }
    
        /// <summary> Read BuildAction from reader </summary>
        public static BuildAction ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new BuildAction();
            result.EntityType = EntityTypeHelper.ReadFrom(reader);
            result.Position = Vec2Int.ReadFrom(reader);
            return result;
        }
    
        /// <summary> Write BuildAction to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write((int) (EntityType));
            Position.WriteTo(writer);
        }
    
        /// <summary> Get string representation of BuildAction </summary>
        public override string ToString() {
            string stringResult = "BuildAction { ";
            stringResult += "EntityType: ";
            stringResult += EntityType.ToString();
            stringResult += ", ";
            stringResult += "Position: ";
            stringResult += Position.ToString();
            stringResult += " }";
            return stringResult;
        }
    }
}

File: generated-code/rust/codecraft/TcpReadWrite/src/model/build_action.rs
use super::*;

/// Build action
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct BuildAction {
    /// Type of an entity to build
    pub entity_type: model::EntityType,
    /// Desired position of new entity
    pub position: Vec2I32,
}

impl trans::Trans for BuildAction {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.entity_type.write_to(writer)?;
        self.position.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let entity_type: model::EntityType = trans::Trans::read_from(reader)?;
        let position: Vec2I32 = trans::Trans::read_from(reader)?;
        Ok(Self {
            entity_type,
            position,
        })
    }
}